File: Advanced Methods of Programming/Lab1C-MPP/Lab1C-MPP/Controller/ServiceParticipant.cs
using Lab1C_MPP.Domain;
using Lab1C_MPP.Repo;
using Lab1C_MPP.Validators;
using System.Collections.Generic;

namespace Lab1C_MPP.Controller
{
    class ServiceParticipant
    {
        private RepoParticipant repoParticipant;
        private Validators.ValidatorParticipant validatorParticipant;

        public ServiceParticipant(RepoParticipant repo, Validators.ValidatorParticipant validator)
        {
            this.repoParticipant = repo;
            this.validatorParticipant = validator;
        }

        public void Save(Participant participant)
        {
            Participant id = this.repoParticipant.FindOne(participant.GetId());
            if (id == null)
                try
                {
                    this.validatorParticipant.Validate(participant);
                    this.repoParticipant.Save(participant);
                }
                catch (ValidationException e)
                {
                    throw new ValidationException(e.Message);
                }
            else throw new RepositoryException("Participantul exista deja");
        }

        public IEnumerable<Participant> FindAll()
        {
            return this.repoParticipant.FindAll();
        }

        public Participant FindOne(int integer)
        {
            Participant participant = this.FindOne(integer);
            if (participant == null)
                throw new RepositoryException("Participant inexistent!");
            else return participant;
        }

        public int Size()
        {
            return this.repoParticipant.Size();
        }

        public int MaxId()
        {
            return this.repoParticipant.MaxId();
        }
    }
}

File: Advanced Methods of Programming/Laborator1JAVA/src/main/java/service/ServiceParticipant.java
package service;

import domain.Participant;
import repo.RepoException;
import repo.RepoParticipant;
import validatori.ValidationException;
import validatori.ValidatorParticipant;

public class ServiceParticipant {
    private RepoParticipant repoParticipant;
    private ValidatorParticipant validatorParticipant;

    public ServiceParticipant(RepoParticipant repoParticipant,ValidatorParticipant validatorParticipant){
        this.repoParticipant=repoParticipant;
        this.validatorParticipant=validatorParticipant;
    }

    public void save(Participant object) throws ValidationException {
        Participant id=this.repoParticipant.findOne(object.getId());
        if(id==null)
            try {
                this.validatorParticipant.validate(object);
                this.repoParticipant.save(object);
            } catch (ValidationException e) {
                throw new ValidationException(e.getMessage());
            }
        else throw new RepoException("Participantul exista deja");
    }

    public Iterable<Participant> findAll() { return this.repoParticipant.findAll(); }

    public Participant findOne(Integer integer) {
        Participant participant=this.findOne(integer);
        if(participant==null)
            throw new RepoException("Participant inexistent!");
        else return participant;
    }

    public int size(){ return this.repoParticipant.size(); }

    public Participant findByNameAndAge(String name, int age){
        Participant participant=this.findByNameAndAge(name,age);
        if(participant==null)
            throw new RepoException("Participant inexistent!");
        else return participant;
    }
}
